### PR TITLE
BUILD: Switch to Meson for building

### PR DESCRIPTION
**Pull request by pastewka**
Switched the build system from setuptools to Meson. Added `pyproject.toml`. Removed `setup.py`.

### CODE REVIEW

**Comment by pastewka** [comment]
Hi @jameskermode @prs513rosewood - this seems to work, with the exception of Windows that complains: "Need python for 64-bit, but found 32-bit". @jameskermode - have you seen this before?

**Comment by jameskermode** [comment]
> Hi @jameskermode @prs513rosewood - this seems to work, with the exception of Windows that complains: "Need python for 64-bit, but found 32-bit". @jameskermode - have you seen this before?

I think one needs to setup the development environment twice, once for 32-bit and again for 64-bit. See this example: https://cibuildwheel.readthedocs.io/en/1.x/cpp_standards/#windows-and-python-27 (although it says it's for Py 2.7, I think the dev setup example is still applicable for us.

File: .github/workflows/build-wheels.yml
Comment (jameskermode) [inline .github/workflows/build-wheels.yml:42]: Is this just a temporary change while debugging? I have found that manylinux2010 wheels are still needed on older HPC clusters, so I'd like to include them if possible

**Comment by jameskermode** [comment]
Latest Windows build error may be related to https://stackoverflow.com/questions/61995649/link-fatal-error-lnk1104-cannot-open-file-python37-d-lib

File: .github/workflows/build-wheels.yml
Comment (jameskermode) [inline .github/workflows/build-wheels.yml:54]: Any reason for this change? I did it the other way to have smaller wheels to download.

File: .github/workflows/build-wheels.yml
Comment (pastewka) [reply .github/workflows/build-wheels.yml:42]: This is a temporary change. These lines don't work on Windows, because running the build through bash does not work. Will need to figure out how to do this with a build matrix.

Comment (jameskermode) [reply .github/workflows/build-wheels.yml:42]: Ok. I thought I'd got past that hurdle by setting `shell: bash` in the workflow step.

**Comment by pastewka** [comment]
I switched to 32-bit devel on Windows not it's complain:
```
LINK : fatal error LNK1104: cannot open file 'python37_d.lib
```
although I explicitly specified Python 3.10.

**Comment by jameskermode** [comment]
Probably getting the build to work on a physical Windows machine at least once would be faster than debugging this via CI only. Unfortunately I don't have access to one; tried to setup a VM but now I'm on Apple Silicon that is not straightforward.

**Comment by pastewka** [comment]
> Probably getting the build to work on a physical Windows machine at least once would be faster than debugging this via CI only. Unfortunately I don't have access to one; tried to setup a VM but now I'm on Apple Silicon that is not straightforward.

I can try to that. I have a working windows VM, but will need to figure out how to get MSVC installed.

File: .github/workflows/build-wheels.yml
Comment (pastewka) [reply .github/workflows/build-wheels.yml]: Yes, but that enables bash on Windows - which has this problem: https://github.com/ilammy/msvc-dev-cmd#name-conflicts-with-shell-bash

File: .github/workflows/build-wheels.yml
Comment (pastewka) [reply .github/workflows/build-wheels.yml:54]: It uses the same name for x86 and arm wheels. Will need to fix and revert.

**Comment by pastewka** [comment]
Hi @jameskermode - it looks like building the wheel works now. However, it seems to only build for "manylinux", not sure how transferable that is between the manylinux versions. I also did not test if the wheels actually work.

**Comment by pastewka** [comment]
Sounds good, thanks!

**Comment by jameskermode** [comment]
Wheels install fine and passing tests on manylinux (although I only checked Python 3.8 so far).

Apple Silicon doesn't work, I think just due to the arm64 wheel being overwritten by the x86_64 one which seems to have the same filename. Will see if I can fix that.

Have requested a Windows test via https://github.com/ACEsuit/mace/issues/52

**Comment by jameskermode** [comment]
> Apple Silicon doesn't work, I think just due to the arm64 wheel being overwritten by the x86_64 one which seems to have the same filename. Will see if I can fix that.

Simply ensuring the artefact has the architecture in the suffix is not sufficient. Looks like it's not croos-compiling correctly for arm64. Will check more carefully soon.

**Comment by pastewka** [comment]
We should just copy what scipy does. They have a specific script for macOS ARM builds, see here: https://github.com/scipy/scipy/blob/main/.github/workflows/wheels.yml

**Comment by pastewka** [comment]
(Minus the Fortran business)

**Comment by pastewka** [comment]
Hi @jameskermode - I've tried to adopt the ARM cross-compile from `scipy`, but it creates a wheel `matscipy-0.8.0.rc1.dev88+g8bd0565-cp38-cp38-macosx_10_16_x86_64.whl` which looks like X86 to me. That said, the file extension may just be wrong.

**Comment by jameskermode** [comment]
Do you know the Github CI ssh debug trick? Can add the following after a failed step then ssh and poke around:

```yaml
      - name: Setup tmate session
        if: failure()
        uses: mxschmitt/action-tmate@v3
        timeout-minutes: 120
```

**Comment by pastewka** [comment]
No, didn't know this. Looks great!

**Comment by prs513rosewood** [comment]
@pastewka  It looks like running `python3 -m pip install .` installs two separate packages: one `_matscipy.<cpython>.so` and one `matscipy/...`. I think the compiled `_matscipy` package should be installed in `matscipy/` as a submodule, that's the common practice in python packages with compiled parts. I can make the change if you think it's ok.

**Comment by jameskermode** [comment]
I agree it would an improvement to move `_matscipy` into `matscipy._matscipy`. That's what we do in quippy too.

**Comment by pastewka** [comment]
I have no objections to moving this. I was always under the impression that most of the binary modules resides in the main package directory, though.

**Comment by prs513rosewood** [comment]
Seems that the tests are failing because pip still installs `_matscipy.so` as a standalone package despite my changes. I've had that happen on my machine and fixed it with a newer pip version.

**Comment by prs513rosewood** [comment]
I've added a simple check to cibuildwheel that makes sure both the package and the compiled extension can be imported. Looks like it's working for every wheel, and I've separately tested the windows wheels on bare metal. If the macos wheels work the PR can probably be merged.

**Comment by jameskermode** [comment]
thank you @prs513rosewood! Last time I checked there was still an issue with the Mac arm64 wheel, I'll test that again now.

**Comment by jameskermode** [comment]
MacOS X arm64 wheel still doesn't install. The `.so` file inside is arm64, but both the metadata filename say its x86_64. It would also be nice to have arm64 wheels for other Python versions (3.8, 3.9, 3.10, 3.11 would be consistent with the other platforms). However, to make progress shall we decouple things by removing arm64 support from this PR and working on it in a new one, so as not to hold up releasing Linux and Windows wheels to PyPI?

**Comment by jameskermode** [comment]
> I've added a simple check to cibuildwheel that makes sure both the package and the compiled extension can be imported. Looks like it's working for every wheel, and I've separately tested the windows wheels on bare metal. If the macos wheels work the PR can probably be merged.

Thanks, good idea - but unfortunately this won't work for arm64 since they are cross-compiled on x86_64 (until GitHub get Apple Silicon runners which I did hear about somewhere as a future) plan. Another reason to work on arm64 in a new PR.

**Review by jameskermode — APPROVED** [review]

**Comment by prs513rosewood** [comment]
The cibuildwheel test actually runs only for wheels built by cibuildwheel. I'm ok with pushing the arm builds to a separate PR, but before merging I'll make a quick change: Numpy [recommands](https://numpy.org/devdocs/user/troubleshooting-importerror.html#c-api-incompatibility) using `oldest-supported-numpy` as a build requirement instead of regular Numpy, to avoid backwards incompatibility (i.e. if the build system pulls a version of numpy at compile time more recent than the one available at runtime).

**Comment by jameskermode** [comment]
I agree with using oldest-supported-numpy - we also do it with f90wrap and quippy wheels (and I thought I'd done it here too but must have got lost in all the edits in this branch

**Pull request merged**
